### ISSUE TITLE
Tune i/o timeout for network switching on mobile devices

### ISSUE THREAD

**Issue by madeye** [xtaci/kcptun#93]
On mobile devices, the underlying network would switch between wifi and 3g/4g frequently. The problem is that kcptun needs to wait for up to one minute to detect the change.

Is it possible to tune the timeout parameters of yamux to improve this issue?

**Comment by xtaci** [comment]
for android, you have to do this by detecting network status change, for eg:

http://stackoverflow.com/questions/3767591/check-intent-internet-connection

and establish reconnection to kcptun server.

changing ping interval doesn't help in this scenario, because short interval will lead to false positive.

**Comment by madeye** [comment]
Cool! Thanks for the input.

It works quite well: https://github.com/shadowsocks/shadowsocks-android/commit/66ceed8607b7e1d1288ec089585fb590b7a1a595

**Comment by Mygod** [comment]
There should be a better way to do this. This is battery draining.

**Comment by xtaci** [comment]
@Mygod  maybe you can use 2 hops , eg:
mobile -> aliyun -> vps

**Comment by wxyzh** [comment]
这是要集成到ss-android

xtaci <notifications@github.com>于2016年8月6日周六 18:26写道：

**Comment by Mygod** [comment]
Could you reopen this issue? I'm still kind of not satisfied with the current solution.

**Comment by madeye** [comment]
Found the default value of ConnectionWriteTimeout is 10s:

https://github.com/hashicorp/yamux/blob/master/mux.go#L43

Is it possibile that kcptun uses the same config?

**Comment by xtaci** [comment]
it's possible.   any more suggestions, i'm preparing a new release
@madeye

**Comment by madeye** [comment]
Nope. But, I'm thinking of connection restore like [what Google does in QUIC]:

> A QUIC connection is uniquely identied by a CID (Con-
nection IDentier) at the application layer and not by the
pairs of IP addresses and port number. The rst advantage
is that, since CIDs are not based on IP addresses, any han-
dover between two networks can be transparently handled
by QUIC without needing to re-establish the connection.
Moreover, the CID is useful in the case of NAT unbinding,
when to restore a connection, a new pair of IP addresses
is typically required. Finally with CID, QUIC can provide
native support to multi-path; for example a mobile device
could use all the network connections for the same CID.
It is important to notice that, at the time of this writing,
multi-path is still not implemented in Chrome.

It may require modification of yamux. Or we can figure out other dirty but quick ways to implement it.

[what Google does in QUIC]: http://c3lab.poliba.it/images/3/3b/QUIC_SAC15.pdf

**Comment by madeye** [comment]
For example, we have a CID (8-byte) in each UDP packet. For each packet, we try to restore the KCP connection from that CID first.

**Comment by xtaci** [comment]
you don't need to change yamux, all you need to do , is changing the identifier of the client:

https://github.com/xtaci/kcp-go/blob/master/sess.go#L673

by replacing address:port tuple with CID

in order to do this, we must create a session initiation process:
1.  client -> server : connect with some HELO
2. server -> client : CID
3. CID is applied to kcp conversation id.

**Comment by xtaci** [comment]
but , step 2 should pay attention to DDoS attack, but flooding with HELO packet

**Comment by madeye** [comment]
Cool! It looks doable.

We may use a small timeout for step 2. And QUIC should suffers from the same problem.

**Comment by xtaci** [comment]
A feasible attacking method for STEP-2 is REPLAY-ATTACK.

some timestamp must be added to the HELO STEP.1.  for example UNIXTIME(), the packet is dropped if the time is  now-60 < ts  < now+60

**Comment by xtaci** [comment]
another problem is, on the client side, you have to separate struct KCP with struct UDPSession to establish reconnection, it's a big change for the library,  maybe you can modify this on your branch.

**Comment by madeye** [comment]
Maybe we can assume connection restore only happens on the server side?

**Comment by xtaci** [comment]
a new ```UDPSession``` losts information on ```struct KCP``` eg, sn, acklist..  a reconnection would not success if ```struct KCP``` cannot survive across  multiple ```UDPSession```(a.k.a UDP connection) close.

The best way to do this is a ```Reconnection-Timeout``` implemented inside UDPSession.

**Comment by xtaci** [comment]
there is another way to improve network efficiency:

***multiple udp connections fusion*** : we can combine several connections into  a single KCP connection, if we un-bind a kcp object from UDPSession.

any of CID & fusion will introduce massive changes to the library, i think, compatibility is a challenge.  I think a standalone repository would be apropriate.

**Comment by madeye** [comment]
What about kcp2-go? :laughing:

**Comment by xtaci** [comment]
no matter what, don't say it's kcptun to make a distinction between the original version.

I have to keep compatibility with game server usage scenario; for games, black technologies like fusion and reconnection are not required, ```responsiveness , simplicity and  backward-compatibility``` is a must to ```kcp-go```.

**Comment by madeye** [comment]
Fogot to mention. I have another proposal about upnp. It won't fix this issue but could help a lot by adding a port mapping of the home routers.

**Comment by xtaci** [comment]
@madeye cool.

another problem for CID is, unexpected server redirection, suppose
1. a connection established with : ```Alice   ->  Bob```
2. Eve captured a packet from Alice, and replay with : ```Eve ---(captured)---> Bob``` with same CID
3. Bob thought that Alice has changed IP, so Bob ```switched the destination addr:port``` to Eve's host.

Eve success hijacked the connection!

**Comment by madeye** [comment]
Good point. it means CID should be encrypted all the time. All the application over kcp should handle the authentication (like OTA in shadowsocks).

**Comment by xtaci** [comment]
I don't think there exists a simple solution to this problem. OTA suffers from REPLAY-ATTACK too.

**Comment by madeye** [comment]
I did some dirty modifications and it seems to work very well. https://github.com/madeye/kcp-go

**Comment by xtaci** [comment]
@madeye I've read the code, but I think it won't work.

if the client IP has changed.  the ```UDPSession.remote``` of server haven't change.

https://github.com/madeye/kcp-go/blob/master/sess.go#L722

the ```remote addr``` of UDPSession has binded ONLY ONCE.

**Comment by madeye** [comment]
https://github.com/madeye/kcp-go/blob/master/sess.go#L731

I changed the remote here. Not sure if it works.

**Comment by xtaci** [comment]
I see, setting remote alone has data races, you can try ```go test -race -run Speed``` to see the races. Mutexes are required to for ```remote addr```

And I have some suggestions for this.
1. SID should be protected by encryption
2. A timestamp(32bit ms) should be added to the header, and only allows a very limited range of remote switching (like < 200ms, <RTO maybe) time range.

PS. QUIC security considerattions
https://docs.google.com/document/d/1g5nIXAIkN_Y-7XJW5K45IblHd_L2f5LTaDUDwvZ5L6g/edit#    
https://www.ietf.org/proceedings/96/slides/slides-96-irtfopen-1.pdf

**Comment by madeye** [comment]
Thanks for the suggestions!

@Mygod I did some experiments by redialing the PPPoE on my router, No more connection loss now. You may want to try the `kcp-with-sid` branch. Remeber to also update your server.

**Comment by xtaci** [comment]
some quotes from QUIC on how to prevent session hijacking:
> The IP address spoofing problem is handled by issuing the client, on demand, a “source-address token”. This is an opaque byte string from the client's point of view. From the server's point of view it's an authenticated-encryption block (e.g. AES-GCM) that contains, at least, the client's IP address and a timestamp by the server. The server will only send a source address token for a given IP to that IP. Receipt of the token by the client is taken as proof of ownership of the IP address in the same way that receipt of a TCP sequence number is.

>Clients can include the source address token in future requests in order to demonstrate ownership of their source IP address. If the client has moved IP addresses, the token is too old, or the client doesn’t have a token, then the server may reject the connection and return a fresh token to the client. But if the client has remained on the same IP address then it can reuse a source-address token to avoid the round trip needed to obtain a fresh one.


actually the addr:port tuple is encapsulated in the UDP packet .

**Comment by madeye** [comment]
It's not very clear to me that if QUIC issues token per source IP, how it recovers the connection when the source addr:port changes.

>If the client has **moved IP addresses**, the token is too old, or the client doesn’t have a token, then the server **may reject the connection** and return **a fresh token** to the client

**Comment by xtaci** [comment]
the server will reject with a fresh token, containing the new ip address and it's encrypted.
so Eve cannot decrypt this information, so it's impossible to create a new &  legal packet to server.

TLS is implemented at the packet level, it guarantees a packet cannot be forged.

**Comment by Mygod** [comment]
Imo this issue isn't resolved yet since `sid` option is removed now.

**Comment by madeye** [comment]
Let's close this issue for now. At least, kcptun works well on Android now.

**Comment by xtaci** [comment]
@madeye Can you modify your https://github.com/shadowsocks/shadowsocks-android/releases/tag/v3.1.0 to reference to https://github.com/xtaci/kcptun/releases/tag/v20160904, it includes tiny modifications.

**Comment by madeye** [comment]
@xtaci Done.
